### PR TITLE
External wallet: connection

### PR DESCRIPTION
**Pull request by NotEternal**
## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/swaponline/MultiCurrencyWallet/blob/master/docs/CONTRIBUTING.md) guide
- [x] Good naming (as clear and simple as possible)
- [ ] Correct behavior if external API endpoints are down, return 404, 504 (or no answer), 401 errors (ddos simulation)
- [x] I tested desktop/mobile resolution
- [x] I tested light/dark theme
- [ ] I tested different languages
- [ ] I checked the functionality once again (**AFFECT MONEY**)
- [x] I checked the work on the Testnet
- [x] I checked the work on the Mainnet
- [ ] I checked the work in the plugin
- [x] I checked the **PR** once again

## Tests

Please start auto tests as follows:

- add a label <ins>swap test</ins> to start swap tests
- add a label <ins>withdraw test</ins> to start withdraw tests

You can skip these tests if you completely sure that your changes aren't related to this functional

### Original issue
#4765

### Perhaps this is related
#4723
#4740

### Video / screenshot proof
image
---
image

### CODE REVIEW

**Comment by developer24-exe** [comment]
builded https://testnet4775.swaponline.site ;

**Comment by developer24-exe** [comment]
commit 8ce9039dd rebuilded https://testnet4775.swaponline.site https://mainnet4775.swaponline.site

File: src/common/web3connect/index.ts [hunk 1]
```diff
@@ -9,11 +9,11 @@ import { isMobile } from 'react-device-detect'
 import config from 'app-config'
 
 export default class Web3Connect extends EventEmitter {
-  _cachedProvider = null
-  _cachedProviderName = null
-  _cachedChainId = null
+  _cachedProvider: IEtheriumProvider | null = null
+  _cachedWeb3: IEtheriumProvider | null = null
```
Comment (shendel) [inline src/common/web3connect/index.ts:13]: Чем _cachedProvider не устроил?
Дубль объекта

**Review by shendel — CHANGES_REQUESTED** [review]

File: src/common/web3connect/index.ts [hunk 2]
```diff
@@ -197,41 +193,40 @@ export default class Web3Connect extends EventEmitter {
...
-
-    // @ToDo - Hard fix walletconnect
-    // https://github.com/WalletConnect/walletconnect-monorepo/issues/384
-    if (window) {
-      window.send = (e,t) => {
-        return _web3provider.send(e,t)
+    if (this._cachedProvider) {
+      this._cachedAddress = await this._cachedProvider.getAccount()
+      this._cachedChainId = await this._cachedProvider.getChainId()
+      const _web3provider = await this._cachedProvider.getProvider()
```
Comment (shendel) [inline src/common/web3connect/index.ts:199]: Сдесь уже объект Web3, у которого есть поле 'eth'

File: src/common/web3connect/index.ts [hunk 3]
```diff
@@ -197,41 +193,40 @@ export default class Web3Connect extends EventEmitter {
...
       }
-    }
 
-    //@ts-ignore: strictNullChecks
-    this._cachedWeb3 = new Web3(
-      _web3provider
-    )
-    //@ts-ignore: strictNullChecks
-    this._cachedWeb3.isMetamask = true
+      this._cachedWeb3 = new Web3(_web3provider)
```
Comment (shendel) [inline src/common/web3connect/index.ts:210]: Создание нового Web3 объекта из объекта Web3

File: src/common/web3connect/providers/InjectedProvider.ts [hunk 4]
```diff
@@ -15,18 +15,14 @@ export default class InjectedProvider extends InjectedConnector {
   }
 
   async isConnected() {
+    // This is opera
     if ((!!window.opr && !!window.opr.addons) || !!window.opera || navigator.userAgent.indexOf(' OPR/') >= 0) {
-      // This is opera
-      //@ts-ignore: strictNullChecks
-      await window.ethereum.enable()
```
Comment (shendel) [inline src/common/web3connect/providers/InjectedProvider.ts:21]: У оперы по документации надо явный enable, если один раз дали разрешение, окон не будет, но и без enable работать не будет

**Review by shendel — DISMISSED** [review]
Тут только UI?

**Comment by developer24-exe** [comment]
commit 8b8dcfd53 rebuilded https://testnet4775.swaponline.site https://mainnet4775.swaponline.site

**Comment by developer24-exe** [comment]
commit e1f973eca rebuilded https://testnet4775.swaponline.site https://mainnet4775.swaponline.site

**Comment by developer24-exe** [comment]
commit 8caebb029 rebuilded https://testnet4775.swaponline.site https://mainnet4775.swaponline.site

**Comment by developer24-exe** [comment]
commit 3520c4c25 rebuilded https://testnet4775.swaponline.site https://mainnet4775.swaponline.site

**Comment by developer24-exe** [comment]
commit 71412f809 rebuilded https://testnet4775.swaponline.site https://mainnet4775.swaponline.site

**Comment by developer24-exe** [comment]
commit 46d0fe245 rebuilded https://testnet4775.swaponline.site https://mainnet4775.swaponline.site

**Review by noxonsu — APPROVED** [review]

**Comment by noxonsu** [comment]
подключился с матика, с китайского валета, все ок

**Review by VitaliyShulik — APPROVED** [review]

**Comment by developer24-exe** [comment]
commit f52b1dd83 rebuilded https://testnet4775.swaponline.site https://mainnet4775.swaponline.site

**Pull request merged**